File: src/main/kotlin/ir/syrent/velocityonlinetime/VelocityOnlineTime.java
package ir.syrent.velocityonlinetime;

import com.google.inject.Inject;
import com.velocitypowered.api.command.CommandMeta;
import com.velocitypowered.api.event.Subscribe;
import com.velocitypowered.api.event.proxy.ProxyInitializeEvent;
import com.velocitypowered.api.plugin.Plugin;
import com.velocitypowered.api.proxy.ProxyServer;
import ir.syrent.velocityonlinetime.controller.DiscordController;
import ir.syrent.velocityonlinetime.database.MySQL;
import ir.syrent.velocityonlinetime.database.SQL;
import ir.syrent.velocityonlinetime.listener.DisconnectListener;
import ir.syrent.velocityonlinetime.listener.SeverConnectedListener;
import ir.syrent.velocityonlinetime.storage.Settings;
import org.slf4j.Logger;

@Plugin(
        id = "velocityonlinetime",
        name = "VelocityOnlineTime",
        version = BuildConstants.VERSION,
        description = "OnlineTime plugin for velocity servers",
        url = "syrent.ir",
        authors = {"Syrent"}
)
public class VelocityOnlineTime {

    public SQL sql;
    public DiscordController discordController;
    public final ProxyServer server;
    public final Logger logger;

    @Inject
    public VelocityOnlineTime(ProxyServer server, Logger logger) {
        this.server = server;
        this.logger = logger;
    }

    @Subscribe
    public void onProxyInitialization(ProxyInitializeEvent event) {
        initializeMySQL();
        Settings.INSTANCE.load();
        discordController = new DiscordController(this);

        registerListeners();
        registerCommands();

    }

    public void initializeMySQL() {
        // TODO: Read data from yaml file
        String host = Settings.INSTANCE.getHost();
        String database = Settings.INSTANCE.getDatabase();
        String username = Settings.INSTANCE.getUsername();
        String password = Settings.INSTANCE.getPassword();
        int port = Settings.INSTANCE.getPort();

        sql = new MySQL(this, host, port, database, username, password);

        try {
            logger.info("Connecting to sql...");
            sql.openConnection();
            sql.createTable();
            logger.info("Connected to sql.");
        } catch (Exception e) {
            logger.error("Error while connecting to sql. Please send error to plugin developer: " + e.getMessage());
            e.printStackTrace();
        }
    }

    public void registerListeners() {
        server.getEventManager().register(this, new SeverConnectedListener(this));
        server.getEventManager().register(this, new DisconnectListener(this));
    }

    public void registerCommands() {
        server.getCommandManager().register("onlinetime", new OnlineTimeCommand(this, discordController));
        CommandMeta meta = server.getCommandManager().metaBuilder("onlinetime").aliases("onlinetime", "pt", "ot").build();
        server.getCommandManager().register(meta, new OnlineTimeCommand(this, discordController));
    }
}

File: src/main/java/ir/syrent/velocityonlinetime/velocity/VelocityOnlineTime.kt
package ir.syrent.velocityonlinetime.velocity

import com.google.inject.Inject
import com.velocitypowered.api.event.Subscribe
import com.velocitypowered.api.event.proxy.ProxyInitializeEvent
import com.velocitypowered.api.plugin.Plugin
import com.velocitypowered.api.plugin.annotation.DataDirectory
import com.velocitypowered.api.proxy.ProxyServer
import com.velocitypowered.api.proxy.messages.ChannelIdentifier
import com.velocitypowered.api.proxy.messages.MinecraftChannelIdentifier
import ir.syrent.velocityonlinetime.BuildConstants
import ir.syrent.velocityonlinetime.velocity.command.OnlineTimeCommand
import ir.syrent.velocityonlinetime.velocity.listener.DisconnectListener
import ir.syrent.velocityonlinetime.velocity.listener.PluginMessageListener
import ir.syrent.velocityonlinetime.velocity.listener.SeverConnectedListener
import ir.syrent.velocityonlinetime.velocity.storage.Database
import ir.syrent.velocityonlinetime.velocity.storage.Settings
import me.mohamad82.ruom.VRUoMPlugin
import org.slf4j.Logger
import java.nio.file.Path

/**
 * Velocity version of BungeeOnlineTime (https://github.com/R3fleXi0n/BungeeOnlineTime) with extra features
 * like weekly rewards, per server online time, discord integration, staff monitoring,  and more.
 * You only need to install plugin in Velocity plugins folder,
 * But if you want to use plugin placeholders using PlaceholderAPI you need to also install plugin in back-end server.
 */
@Plugin(
    id = "velocityonlinetime",
    name = "VelocityOnlineTime",
    version = BuildConstants.VERSION,
    description = "OnlineTime plugin for velocity servers",
    url = "syrent.ir",
    authors = ["Syrent"]
)
class VelocityOnlineTime @Inject constructor(
    server: ProxyServer,
    logger: Logger,
    @DataDirectory dataDirectory: Path
) : VRUoMPlugin(server, logger, dataDirectory) {

    @Subscribe
    fun onProxyInitialization(event: ProxyInitializeEvent) {
        instance = this
        getDataDirectory().toFile().mkdir()
        getServer().channelRegistrar.register(VELOCITYONLINETIME_CHANNEL)
        Settings
        Database
        registerListeners()
        registerCommands()
    }

    /**
     * Plugin starts a MilliCounter $[me.mohamad82.ruom.utils.MilliCounter]  whenever
     * a player joins the server and ends it when the player disconnects or changes the server.
     * Whenever the player disconnect or change the server, the plugin will save his online time (LeaveTime - JoinTime) to database.
     */
    fun registerListeners() {
        SeverConnectedListener(this)
        DisconnectListener(this)
        PluginMessageListener(this)
    }

    fun registerCommands() {
        OnlineTimeCommand(this)
    }

    companion object {
        var instance: VelocityOnlineTime? = null
            private set

        /**
         * Create new minecraft channel identifier
         * The name should be same name that used in [ir.syrent.velocityonlinetime.spigot.VelocityOnlineTimeSpigot]
         */
        val VELOCITYONLINETIME_CHANNEL: ChannelIdentifier =
            MinecraftChannelIdentifier.create("velocityonlinetime", "main")
    }
}